### ISSUE TITLE
Pasting links in Gutenberg

### ISSUE THREAD

**Issue by amamujee** [WordPress/gutenberg#21789]
**Feature Request:**It would be cool if users had the option to select if pasting in a link is just a link vs. a link with some kind of preview. Here is a screen recording of how the twitter link pasting works on WP.com vs. link pasting.

**Solution options:**
1) Link should not be pasted as text, but autolinked (should happen regardless)
2) Give users the option to past as a link or as a preview
3) A link pasted behaves like the Twitter embed

I think the middle option makes the most sense, and the first should happen regardless. I could see some value in making it a separate block if we wanted the experience to be 3)

**Comment by mapk** [comment]
We need some design around this. There are a couple issues related to this concept and it seems to become increasingly popular.

Related: 
https://github.com/WordPress/gutenberg/issues/15102
https://github.com/WordPress/gutenberg/issues/18653

**Comment by paaljoachim** [comment]
Associated:
https://github.com/WordPress/gutenberg/issues/21029

PR:
https://github.com/WordPress/gutenberg/pull/17413

**Comment by lancewillett** [comment]
Related: https://github.com/WordPress/gutenberg/issues/20442 

Where the title from target page is created as the link text — like Google Docs.

**Comment by karmatosed** [comment]
I'm going to dive into this and see how we can move it along a little. I am going to try and see if I understand the proposal first, to make sure. If I have got that wrong there are a lot of issues linked here, so I can course-correct with feedback.

> Link should not be pasted as text, but autolinked (should happen regardless)

I agree with the idea that a link should auto-link by default, that just makes sense. The key raised here is choice, having the ability to choose either I think is where we can refine the flow. There seem 2 options here:

- Add a step to ask if want to embed or just have text link.
- Try and assume the best result, then offer an alternative.

I haven't included this option, but can explore if desired:

> A link pasted behaves like the Twitter embed

My feeling was to embed all links could end up feeling very distracting.

## Offering a step in between

<img width="752" alt="5" src="https://user-images.githubusercontent.com/253067/81696343-af2b3280-945b-11ea-8789-4fb6357a60cb.png">

This is a rough mockup just as a visual to get feedback on. In this, it would only do this if could embed, a link that was just a text-link wouldn't offer this.

## Offering option to convert to text link only
Here you could have 2 optional locations, one in the toolbar (icon would need work) and one in the ellipsis menu.

<img width="711" alt="Frame 2" src="https://user-images.githubusercontent.com/253067/81695239-455e5900-945a-11ea-9c8c-fa1e19e12651.png">

<img width="631" alt="Frame 1" src="https://user-images.githubusercontent.com/253067/81695220-3f687800-945a-11ea-9e36-19228c3c1079.png">

## Recommendation
My own feeling is avoiding the extra step probably is sensible in this case, but giving an option to change is good. Having it load then offering a modal or 'splash' that says 'continue or make as link' could also be another option, again though I wonder if it's adding a hurdle. Ultimately, that is a decision we need to take as to what is the default, but being able to adjust feels right.

## Additional steps forward
The above is just one soothing of the experience. Let's do more by bringing in some additional tickets linked here.

- Automatically formatting a new hyperlink with the title text from the link destination: #20442
- Always give a choice to convert to link: https://github.com/WordPress/gutenberg/issues/18653

## Wrapping up into next steps
What would be great would be to get feedback on the above and then move the linked issues into development along with this one. I'm going to put the feedback label on so we can start that journey.

**Comment by paaljoachim** [comment]
In Gutenberg 8.1 one can choose to transform any embed into a Paragraph Block. See this comment: https://github.com/WordPress/gutenberg/pull/17413#issuecomment-627371167

**Comment by mapk** [comment]
> My own feeling is avoiding the extra step probably is sensible in this case, but giving an option to change is good.

I agree with this. Just as #17413 provides an option after the embed happens, a similar option can also work here as you've mocked up above. 

Of the options, the most clearly understandable is the ellipses dropdown that you have with a "Convert to link" option. I'd suggest keeping it Sentence case and then converting the other options to Sentence case in another PR.

**Comment by karmatosed** [comment]
@paaljoachim that's one option, for me though transforming feels a bit hidden, I don't know I guess you could argue that it's just as hidden in the ellipsis. I wonder though if you would associate paragraph with changing to a link?

**Comment by paaljoachim** [comment]
Hey Mark and Tammie.
The other option which I believe @desaiuditd will work on next is this:
https://github.com/WordPress/gutenberg/issues/21029#issuecomment-619055961

**Comment by desaiuditd** [comment]
> that's one option, for me though transforming feels a bit hidden, I don't know I guess you could argue that it's just as hidden in the ellipsis. I wonder though if you would associate paragraph with changing to a link?

@karmatosed 
Just to be clear, are we thinking to keep both the options? That would be bad UX. Two ways of doing the same thing. Isn't it? Might be confusing and ambiguous to end-users.

**Comment by karmatosed** [comment]
@desaiuditd I wouldn't suggest both, I would suggest just one and it's about deciding which one.

**Comment by paaljoachim** [comment]
Thank you Tammie, for diving into this issue! Inspired by your thoughts as well as thoughts I have about this process here is my feedback.

A possible process of pasting a link.
**Can the link be embedded.**
No. Show the text link.
Yes. Show a placeholder screen.

1a
**Add the placeholder right after the user has pasted the link that can be embedded.**

![New-Embed-Placeholder](https://user-images.githubusercontent.com/5323259/81855604-dbc87280-955f-11ea-990a-22543501c00c.jpg)

The question that shows up.... is it intrusive to create the extra step in adding an Embed URL placeholder screen?

1b
**An option: Embeds are automatically added. User clicks pencil to bring up the placeholder screen.**

![Screen Shot 2020-05-13 at 21 27 22](https://user-images.githubusercontent.com/5323259/81856161-b0925300-9560-11ea-8b91-996de68aeb9d.png)

To change an embed to a text link the user can click the pencil to get access to the placeholder screen to where they can choose to convert the embed to a text link.

2.
Convert to link through the ellipsis menu. Can be done, but I believe we can easily begin to place too many options into this menu.

A familiar method is to use the visible pencil/or another edit icon to create an effect to the selected block.

**Comment by paaljoachim** [comment]
I made an Embed prototype:
https://www.figma.com/proto/OfdH1LSa2MYLUt7iUDRhq7/Embed?node-id=1%3A2&scaling=min-zoom

It uses a placeholder and the 3 dot ellipsis menu.

**Comment by karmatosed** [comment]
I think it would be a good idea to focus in on the problem we're trying to solve and recapping where we are today. I appreciate the mocks, but I do think it's a good spot to check in on objective here.

Currently, we have the option to transform as shown in #17413.

Next steps based on revising this flow would be to bring in the following:

- Offer of choice on the placeholder
- Option to have in block to convert

Now let's try and focus on each of those.

## Placeholder
This would appear only if you paste a link which could embed. I think a simpler approach as I outlined here would be a good start for moving to code.

<img width="752" alt="81696343-af2b3280-945b-11ea-8789-4fb6357a60cb" src="https://user-images.githubusercontent.com/253067/82567697-a3372300-9b75-11ea-8244-52a6318c3cc3.png">

## Option to convert
There are a few proposed spots for this:

- Transforms: merged
- New button: adding new action into toolbar
- Ellipsis
- Hooking into pencil (edit)

I would suggest the ellipsis for me works best here and would like to see that over the transforms, which I am not totally convinced someone will think about. I could be persuaded for a transform and ellipsis, but I really am not sure this would be something you'd go there to discover. Again, happy to discuss that though.

## Next steps
Whilst I appreciate the mocks I do think we can scale this back and see what the effect is on people. We don't need to dive right into a full scale dual button within editing or other options right now.

My feeling is we can based on the following move into development, but I would appreciate some feedback before as I am making some decisions here.

1. We add a placeholder that appears if you paste a link that can also embed.
2. We move transform to link to the ellipsis as show here:

<img width="631" alt="81695220-3f687800-945a-11ea-9e36-19228c3c1079" src="https://user-images.githubusercontent.com/253067/82567759-bb0ea700-9b75-11ea-9808-105a78b2a5be.png">

I am closing #8653 so we can focus on this issue and get a solution. After that, it can be iterated. Let's see this as a step and then get feedback.

**Comment by paaljoachim** [comment]
As you mentioned we have one way that is already merged which is to transform an embed from inside the standard Transform menu to a paragraph block (is not easy to be aware of but uses an existing method). 

1.
Pasting a Youtube URL or any other link that automatically embeds, a placeholder shows up asking the question "Add embed" or "Add text link". It gives more control to how the user wants to show any type of link that automatically shows an embed.

I think the easiest way would be to add a general placeholder box for all embeds. Creating an embed box for each embed will likely be a lot of work making it too complex.  

2. 
Moving (not duplicating) the transform into the ellipsis menu seems to me as stepping away from a general usability method of how the Transform menu is used today. Blocks in general have a method to transform into other kinds of block through the existing Transform menu (even though it is a bit hidden).

**Comment by karmatosed** [comment]
> I think the easiest way would be to add a general placeholder box for all embeds. Creating an embed box for each embed will likely be a lot of work making it too complex.

Just to be clear my note is to add a link and copy change not to change the boxes for each placeholder, that then allows us to avoid a placeholder before the embed one.

**Comment by mapk** [comment]
I agree it's frustrating if I just want to paste in a link without an embed. But I feel this way because the option to convert it doesn't exist.

I'd like to provide the user a decisive solution and then offer options if they wish to change it.

This means that if we have an embed option, we just embed it. BUT we allow the user to convert it to a link using the ellipses menu as @karmatosed mocked up. Ultimately, I'd also like to see us allow the ability to paste without formatting, @ellatrix. Like maybe a `Shift + CMD + V` to get an unformatted paste. It wouldn't have it linked, but then the user can just make it a link from the toolbar.

What do y'all think if we started here? If you feel the placeholder is necessary when pasting a link that can be an embed, I'm open to that direction.

**Comment by paaljoachim** [comment]
Thanks for sharing Mark!

I am open for the various ideas.
Placeholder.
Ellipsis menu.
Let's just leave the newly merged transform Embed to Paragraph block where it is. We can always add on additional methods. As each Block have their own transform menu options. 

Having the **Convert to text link** option show up in the ellipsis menu when an embed is selected is very helpful. As the ellipsis menu will likely show relevant options depending on which block is selected. So the option would not be there on clicking blocks that are not embeds.

Sounds like a good move forward to add the option into the ellipsis menu.

**Comment by paaljoachim** [comment]
@karmatosed  
Can we ask @desaiuditd to begin development of adding a **Convert to (text) link** in the Ellipsis menu for Embeds?

**Comment by karmatosed** [comment]
@desaiuditd that's fine, but please also remove the one for transforms, so we don't have both.

**Comment by paaljoachim** [comment]
Hey Tammie. In regards to two methods of transforming an embed. One in the **Transform to** panel next to Group. I see no harm in keeping the already merged transform to Paragraph in the Transform to panel. As it gives the user two different methods to convert an embed. Users will choose different ways. The Transfom to panel is meant as a place to transform the selected block to some other block. Transforming the Embed to a Paragraph block also belongs in the specific panel.

<img width="718" alt="Screen Shot 2020-05-26 at 16 58 16" src="https://user-images.githubusercontent.com/5323259/82916273-1ecc1200-9f72-11ea-95e3-fd25cd9fdc21.png">

Having the "Convert to text link" in the Ellipsis menu.

![Convert-to-Text-link](https://user-images.githubusercontent.com/5323259/82916694-a6198580-9f72-11ea-88a6-9fdec22c9b8d.jpg)

Will likely be the more normal way for most to change an embed to a text link.

In general lets add blocks to the **Transform to** panel. Giving the option for users to in general use this panel to transform any kind of block. This panel becomes a kind of sidewing to any other method. For instance Group is listed in the Transform to panel and in the Ellipsis menu.

**Comment by mapk** [comment]
Initially when I read about using the Transform tool to transform the embed to a link, I wasn't sure how that would happen. A "link" isn't really a block. But a Paragraph is! Thanks for the visual, @paaljoachim. This makes a bit more sense to me (but I know Gutenberg), does it make sense to a user? Maybe? Either way, it's probably okay allowing the block to be transformed to a Paragraph regardless because it's utilizing an existing Gutenberg ability.

**Comment by desaiuditd** [comment]
> @desaiuditd that's fine, but please also remove the one for transforms, so we don't have both.

Cool. I'll try to work on this in this week. Will raise a PR.

**Comment by karmatosed** [comment]
Thanks @desaiuditd that would be great. Let's remove transforms and take it from there as far as direction. If it does feel we need more methods then we absolutely can iterate. We need to distil back first though.

**Comment by desaiuditd** [comment]
I started working on this and I saw that Embed block code is written in good ol' React Class based component. And I'm like ... no, I'm not touching that. 😉On serious note, I need to be able to use certain React hooks e.g., `useSelect`, `useDispatch` etc. I could use `withSelect` and `withDispatch` in existing structure.

But then I remembered @mcsf 's comment on #17413 that

> The embeds code base could use an important refresh

So I thought this could be a good opportunity to refactor the Embed block code. At least, converting the edit component into a function component to start with. So that React Hooks can be used within.

https://github.com/WordPress/gutenberg/blob/4857ad58c1241b3d63d21a6880c989b85746c3dc/packages/block-library/src/embed/edit.js#L24-L29

But before I do that, can I get consent from someone senior from the dev community? Any thoughts/opinions?

Of course, I can do that in separate PR, to maintain separation of concerns. One, just for the refactor and second, to actually add the `Convert to Link` option.

**Comment by paaljoachim** [comment]
Hey Udit @desaiuditd 

This would be something we can bring up during the Core Editor meeting on Wednesday. As it is a good place to create some discussions around refactoring the Embed block code.

**Comment by ZebulanStanphill** [comment]
Our coding guidelines [recommend the usage of function components](https://developer.wordpress.org/block-editor/contributors/develop/coding-guidelines/#documenting-wordpresselement-react-components), so I'd say this is a perfect opportunity to refactor the Embed block to use hooks.

> When possible, all components should be implemented as function components, using hooks for managing component lifecycle and state.

**Comment by desaiuditd** [comment]
Great! I’ll open PR soon!

**Comment by jasmussen** [comment]
https://github.com/WordPress/gutenberg/issues/27915 has some valuable additions to this issue, worth a look.

**Comment by jhmonroe** [comment]
Ping! Let's get rid of this annoying auto-embed feature! Me and all my clients who are just trying to paste links don't want or need this auto-embed feature that is clearly meant for bloggers

The link above is interesting too — it should be a preference https://github.com/Automattic/blocks-everywhere/pull/83
